### PR TITLE
feature/PRO-1038/pool-fee-rpc

### PR DESCRIPTION
**Pull request by Janislav**
# Pull Request

Closes: PRO-xxx

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have updated documentation where appropriate.

## Summary

*Please include a succinct description of the purpose and content of the PR. What problem does it solve, and how? Link issues, discussions, other PRs, and anything else that will help the reviewer.*

### CODE REVIEW

**Comment by Janislav** [comment]
Docs are updated here: https://github.com/chainflip-io/chainflip-docs/pull/181

File: state-chain/amm/src/range_orders.rs [hunk 1]
```diff
@@ -176,9 +176,9 @@ pub struct PoolState<LiquidityProvider: Ord> {
 	liquidity_map: BTreeMap<Tick, TickDelta>,
 	positions: BTreeMap<(LiquidityProvider, Tick, Tick), Position>,
 	/// Total fees earned over all time
-	total_fees_earned: SideMap<Amount>,
+	pub(super) total_fees_earned: SideMap<Amount>,
```
Comment (Janislav) [inline state-chain/amm/src/range_orders.rs:179]: Not sure about the return value here. Maybe something scalar would make more sense?

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 2]
```diff
@@ -1157,6 +1157,14 @@ pub struct PoolInfo {
 	/// The fee taken, when range orders are used, from swap inputs that contributes to liquidity
 	/// provider earnings
 	pub range_order_fee_hundredth_pips: u32,
+	/// The total fees earned in this pool by range orders.
+	pub range_order_total_fees_earned: SideMap<Amount>,
```
Comment (Janislav) [inline state-chain/pallets/cf-pools/src/lib.rs:1161]: Not sure about the return value here. Maybe something scalar would make more sense? Same for all the over new added filelds.

**Comment by AlastairHolmes** [comment]
So the fees aren't all taken in a single asset, so HistoricalEarnedFees needs to be a double map.

**Comment by Janislav** [comment]
Okay, interesting. I somehow remembered that we only collect LP fees in USDC. Probably this has changed in the meantime.

**Comment by AlastairHolmes** [comment]
> Okay, interesting. I somehow remembered that we only collect LP fees in USDC. Probably this has changed in the meantime.

thats the network fee

File: state-chain/runtime/src/runtime_apis.rs [hunk 3]
```diff
@@ -75,6 +75,7 @@ pub struct AuctionState {
 pub struct LiquidityProviderInfo {
 	pub refund_addresses: Vec<(ForeignChain, Option<ForeignChainAddress>)>,
 	pub balances: Vec<(Asset, AssetAmount)>,
+	pub earned_fees: BTreeMap<ForeignChain, Vec<(Asset, AssetAmount)>>,
```
Comment (AlastairHolmes) [inline state-chain/runtime/src/runtime_apis.rs:78]: Can you use the new any::AssetMap?

File: state-chain/pallets/cf-lp/src/lib.rs [hunk 4]
```diff
@@ -139,6 +139,11 @@ pub mod pallet {
 	pub type FreeBalances<T: Config> =
 		StorageDoubleMap<_, Twox64Concat, T::AccountId, Identity, Asset, AssetAmount>;
 
+	#[pallet::storage]
+	/// Historical earned fees for an account. Map: AccountId => AssetAmount
+	pub type HistoricalEarnedFees<T: Config> =
```
Comment (AlastairHolmes) [inline state-chain/pallets/cf-lp/src/lib.rs:144]: Actually we could use the AssetMap here too.

Comment (AlastairHolmes) [reply state-chain/pallets/cf-lp/src/lib.rs:144]: And use a StorageMap instead.

File: state-chain/pallets/cf-pools/src/lib.rs (hunk 2 above)
Comment (AlastairHolmes) [reply state-chain/pallets/cf-pools/src/lib.rs:1168]: So we can use SideMap as you have here, and in the other PR of yours we will change it to PairMap (or whatever we call it).

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 5]
```diff
@@ -1953,11 +1966,9 @@ impl<T: Config> Pallet<T> {
 		amount_change: IncreaseOrDecrease<AssetAmount>,
 	) -> DispatchResult {
 		let collected_fees: AssetAmount = collected.fees.try_into()?;
-		T::LpBalance::try_credit_account(
-			lp,
-			asset_pair.assets()[(!order.to_sold_side()).into()],
-			collected_fees,
-		)?;
+		let asset = asset_pair.assets()[(!order.to_sold_side()).into()];
```
Comment (AlastairHolmes) [inline state-chain/pallets/cf-pools/src/lib.rs:1969]: You need to add the recording of fees into `inner_update_range_order` too.

**Review by AlastairHolmes — APPROVED** [review]
Ok, sorry that took a while, once all the comments I've made are addressed, I'm happy for this to be merged.

File: state-chain/pallets/cf-pools/src/lib.rs (hunk 5 above)
Comment (Janislav) [reply state-chain/pallets/cf-pools/src/lib.rs:1969]: I think I do this.

File: state-chain/pallets/cf-pools/src/lib.rs [hunk 6]
```diff
@@ -1520,6 +1528,7 @@ impl<T: Config> Pallet<T> {
 				.try_map(|(asset, collected_fees)| {
 					AssetAmount::try_from(collected_fees).map_err(Into::into).and_then(
 						|collected_fees| {
+							T::LpBalance::record_fees(lp, collected_fees, asset);
```
Comment (Janislav) [inline state-chain/pallets/cf-pools/src/lib.rs:1531]: @AlastairHolmes that's the place where I have to add record the fees for range orders, right?

Comment (AlastairHolmes) [reply state-chain/pallets/cf-pools/src/lib.rs:1531]: Sorry not sure how I could have missed that

**Pull request merged**